lib/puma/thread_pool.rb

```diff
@@ -192,24 +201,39 @@ def <<(work)
     # then the `@todo` array would stay the same size as the reactor works
     # to try to buffer the request. In tha scenario the next call to this
     # method would not block and another request would be added into the reactor
-    # by the server. This would continue until a fully bufferend request
+    # by the server. This would continue until a fully buffered request
     # makes it through the reactor and can then be processed by the thread pool.
-    #
-    # Returns the current number of busy threads, or +nil+ if shutting down.
-    #
-    def wait_until_not_full
+    def wait_until_not_full(sock)
       @mutex.synchronize do
         while true
           return if @shutdown
 
-          # If we can still spin up new threads and there
-          # is work queued that cannot be handled by waiting
-          # threads, then accept more work until we would
-          # spin up the max number of threads.
           busy_threads = @spawned - @waiting + @todo.size
+          threads_available = @max - busy_threads > 0
+
+          timeout = nil
+
+          # If the socket is passed in, do a more dynamic busy checked
+          # to load the workers more evenly.
+          if sock
+            # Accept more work if no threads are busy,
+            # or if extra threads and non-blocking work are available.
+            return if busy_threads.zero? ||
+              (threads_available && IO.select([sock], nil, nil, 0))
+
+            # If threads are available but no work,
+            # check for work again after a short timeout, which is a function
+            # of the available threads.
+            # Otherwise if threads are busy, block without timeout
+            # until a thread finishes its work.
+            timeout = if threads_available
+                        (busy_threads.to_f / @max) * WORK_AVAILABLE_TIMEOUT_FACTOR
+                      end
+          end
+
           return busy_threads if @max > busy_threads
```

Comment:
This line was added by #1648, and makes the behavior of this PR quite different from the one in #1646 (which is based on a commit prior to #1648). Here, the call to `@not_full.wait` will _never_ be reached unless `busy_threads >= @max` (ie, the thread pool is maxxed out with requests), which would be an extremely rare case.

I believe this is the reason this PR has no observable effect in benchmarks.